### DISCUSSION TITLE
One rgw node can run multiple rgw with difference port?

### DISCUSSION THREAD

**Discussion by zhucan** [rook/rook#10397] [Q&A]
Now one rgw node  to one rgw.

**Comment by travisn** [comment]
The same port is used for all the rgw. But this port should just be an internal detail for the pod so there isn't a conflict anyway between pods. Are you seeing conflicts or what is the scenario?

**Comment by zhucan** [comment]
I want to create different rgw on same node with different port,  but I don't know how to do it.

**Reply by travisn** [reply]
Are you using host networking? If not, with pod networking, pods on the same node can use the same port without a problem.

**Comment by zhucan** [comment]
Yes， I'm using the host networking, because of  we want to access the block pool from external client out of the cluster.  we want to create extra rgw on the same node with different port manually.

**Reply by travisn** [reply]
With host networking, there is not currently a way in rook to configure a different port for the rgw.

Another approach is to not use host networking, then just [expose the rgw endpoints with an ingress or similar approach](https://rook.github.io/docs/rook/v1.9/Storage-Configuration/Object-Storage-RGW/object-storage/#access-external-to-the-cluster).

**Comment by zhucan** [comment]
Is there way that deploy block storage with host networking, but for rgw not use host networking?

**Reply by travisn** [reply]
Not currently. I could imagine a setting in the CephObjectStore CR under `spec.gateway` for `hostNetworking: false` that would override the setting in the cluster CR. Want to open an issue for that?

**Reply by zhucan** [reply]
> Not currently. I could imagine a setting in the CephObjectStore CR under `spec.gateway` for `hostNetworking: false` that would override the setting in the cluster CR. Want to open an issue for that?

yes, can I open an issue for that?

**Reply by thotz** [reply]
If understand correctly if host network is enabled then RGW daemon run port specified in objectstorespec otherwise it uses `8080` 
```
func (c *clusterConfig) portString() string {
    var portString string
    
    port := c.store.Spec.Gateway.Port
    if port != 0 {
        if !c.clusterSpec.Network.IsHost() {
            port = rgwPortInternalPort
        }
        portString = fmt.Sprintf("port=%s", strconv.Itoa(int(port)))
    }
...
```
So IMO it is possible to run multiple rgw daemons in node with different ports if host network is enabled. Am I missing something?

**Reply by travisn** [reply]
@thotz I'm assuming the requirement was for the same object store to have multiple rgw daemons per node. But if they all use the same port, that is not possible. @zhucan This is the case, correct?

**Comment by zhucan** [comment]
@travisn yes, if we only have 3 rgw node, we can only create 3 rgw daemons, it is unreasonable.

**Comment by thotz** [comment]
@travisn @zhucan: Now I understand the use case, if you are okay to use different object store CRs for each RGW server. Then this can be set up with a simple multisite configuration. Create a realm/zonegroup/zone configuration using https://rook.io/docs/rook/v1.9/Storage-Configuration/Object-Storage-RGW/ceph-object-multisite/#creating-object-multisite. Now ceph pools for rgw servers will be created when zone CR is created. Now you can create multiple cephobjectstores with different ports using the same zone.  All the data from different rgw servers will be stored in the same ceph pools.

**Reply by travisn** [reply]
That's a pretty complex way to get multiple RGWs with different ports. :) 
@zhucan Why do you need multiple RGWs on the same node for the same object store? If you increase the pod resources, or configure RGW on more nodes, hopefully you could handle the load.

**Reply by richardelling** [reply]
I'm still confused about the request here. But I can share how we do something like this, using the common kubernetes approach to services with a load balancer. We adjust the 'instances' of the CephObjectStore resource. We then create a load balancer (MetalLB, in our case) to point to the rgw pods.
The result is more than one rgw accessible via a single IP adress and port. The topology spreading can be adjusted using normal kubernetes spec. By default, the spread is across nodes first.
The result is slightly better performance and slightly better reliability in exchange for minimal configuration changes or administrative overhead.

**Comment by zhucan** [comment]
@zhucan Because of we will not add node always.
